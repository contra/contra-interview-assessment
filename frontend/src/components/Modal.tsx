/* eslint-disable unicorn/numeric-separators-style */
/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import classNames from 'classnames';
import FocusTrap from 'focus-trap-react';
import React, { useContext, useEffect, useRef } from 'react';
import Button from './Button';

/**
 * TODO:
 * 1) create Modal component and display this on the index page ✅
 * 2) add toggle button on index page to display the modal ✅
 * 3) add prop to show Modal ✅
 * 4) add a way to close Modal ✅
 * 5)
 * 6) add overlay ✅
 * 7) prevent scrolling ✅
 * 8) Handle stacking ✅
 * 8.1) Fix ESC handling on nested dialogs ✅
 * 8.2) Fix scroll disabling on nested dialogs ✅
 * 9) handle focus management and tabbing ✅
 * 10) Allow escape to close the modal ✅
 * 11.1) Fix mobile vertical alignment for modal ✅
 * bonus 1) add modal animation ✅
 * bonus 2) add button animation ✅
 */


type Props = {
    animate?: boolean,
    children: React.ReactNode,
    handleClose: CloseHandler,
    show: boolean,
    title: string,
}

type CloseHandler = () => void;

/**
 * Create overlay component that darkens the background and handles clicks
 */
const Overlay = ({ handleClose, zIndex }: { handleClose: CloseHandler, zIndex: number }) => {

    return <div 
        aria-hidden="true"
        className="absolute top-0 left-0 w-full h-full bg-slate-300/80"
        onClick={handleClose}
        role="button" 
        style={{zIndex}}
        tabIndex={0}
    />;
}

/**
 * We are using a React context to keep track of the nesting level of modals.
 * Then we can use this information to only disable/enable scrolling once for the topmost Modal.
 */
const NestingLevel = React.createContext(0);

const Modal = ({ animate = true, children, handleClose, show, title }: Props) => {

    const nestingLevel = useContext(NestingLevel);
    
    const handleKey = (event: React.KeyboardEvent) => {
        if (event.key === 'Escape' || event.key === 'Esc' || event.keyCode === 27) {
            // without this parent modals will also close
            event.stopPropagation();
            handleClose();
        }
    }

    /**
     * Prevent background from scrolling while the Modal is open
     */
    // eslint-disable-next-line consistent-return
    useEffect(() => {
        if (nestingLevel === 0) {
            if (show) {
                document.body.style.overflow = 'hidden';
            }
            else {
                document.body.style.overflow = 'unset';
            }

            return () => { 
                document.body.style.overflow = 'unset';
            };
        }
    }, [show, nestingLevel]);

    /**
     * Trigger animation to drop the dialog from the top
     */
    const dialogRef = useRef<HTMLDivElement>(null);
    useEffect(() => {
        if (dialogRef.current) {
            dialogRef.current.classList.remove('mt-[-1000px]');
            dialogRef.current.classList.add('mt-0');
        }
    });

    const focusTrapOptions = {
        clickOutsideDeactivates: true,
        // we need escapeDeactivates to be false otherwise nested dialogs will close in the wrong order
        escapeDeactivates: false,
        // we need fallbackFocus for tests to pass
        fallbackFocus: `div[aria-label=${title}]`, 
    }

    const modalStyle = "absolute w-full sm:w-[400px] sm:-ml-[200px] bg-white sm:rounded-xl shadow-xl top-1/4 sm:left-1/2"
    const animation = animate ? "transition-all duration-300 ease-out mt-[-1000px]": "mt-0";

    return show ? (<div className="fixed top-0 left-0 w-full h-full">
        <Overlay handleClose={handleClose} zIndex={1000+nestingLevel*2}/>
        <FocusTrap focusTrapOptions={focusTrapOptions}>
            <div aria-label={title}
                className={classNames(modalStyle, animation)}
                onKeyUp={handleKey} ref={dialogRef} role="dialog"
                style={{zIndex: 1001+nestingLevel*2}} tabIndex={-1}
            >
                <h2 className="text-center bg-violet-700 sm:rounded-t-xl text-white py-1">{title}</h2>
                
                <div className="py-8 px-8 space-y-2 sm:py-4 flex items-center flex-col sm:space-y-0 sm:space-x-6">
                <NestingLevel.Provider value={nestingLevel+1}>
                    {children}
                </NestingLevel.Provider>
                </div>
                <div className="text-center mb-4">
                    <Button onClick={handleClose}>Close</Button>
                </div>
            </div>
        </FocusTrap>
    </div>) : null;
}

export default Modal;