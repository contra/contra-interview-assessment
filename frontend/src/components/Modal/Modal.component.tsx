import { Fragment, MutableRefObject, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import styles from './Modal.module.css';
import KeyboardManager from './keyboardManager';

export default function Modal({ isOpen, onClose = () => {}, children, focusRef }: { isOpen: boolean, onClose?: () => void, children?: JSX.Element, focusRef?: MutableRefObject<HTMLElement | null> }) {
  const modalRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const closeFunction = () => onClose();
    if (isOpen) {
      document.body.classList.add('stopScroll');
      KeyboardManager.register('Escape', closeFunction);
      focusRef?.current?.focus();
    }
    return () => {
      KeyboardManager.unregister('Escape', closeFunction)
      document.body.classList.remove('stopScroll');
    }
  }, [isOpen])

  return isOpen ? ReactDOM.createPortal(
    <Fragment>
      <div className={styles.overlay} onClick={onClose}>
      </div>
      <div className={styles.modalWrapper} ref={modalRef}>
        <div role="dialog" className={styles.modal}>
          {children}
        </div>
      </div>
    </Fragment>,
    document.body
  ) : null;
}
