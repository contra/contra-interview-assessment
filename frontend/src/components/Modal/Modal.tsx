import { Icon } from '@iconify/react';
import React, {
  useEffect,
  useState,
  useMemo,
  type FC,
  useRef,
  useCallback,
} from 'react';
import { createPortal } from 'react-dom';
import { useSpring, animated } from 'react-spring';
import SimpleButton from '../SimpleButton';
import {
  bodyStyle,
  closeButtonStyle,
  headerStyle,
  footerStyle,
  modalWrapperStyle,
  modalStyle,
  overlayVariants,
} from './modal.css';

type PortalProps = {
  children?: React.ReactNode;
};

const Portal: FC<PortalProps> = ({ children }) => {
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
    return () => setMounted(false);
  }, []);

  return mounted ? createPortal(children, document.body) : null;
};

type ModalProps = {
  animate?: boolean;
  cancelButtonText?: string;
  children?: React.ReactNode;
  destroyOnClose?: boolean;
  escapable?: boolean;
  footer?: React.ReactNode;
  header?: React.ReactNode;
  lazy?: boolean;
  maskClosable?: boolean;
  okButtonText?: string;
  onCancel?: () => void;
  onOk?: () => void;
  onOpened?: () => void;
  title?: string;
  visible: boolean;
  width?: string;
};

const Modal: FC<ModalProps> = ({
  animate = true,
  title,
  visible,
  children,
  destroyOnClose = false,
  escapable = true,
  footer,
  header,
  lazy = false,
  maskClosable = true,
  okButtonText = `OK`,
  cancelButtonText = `Cancel`,
  onOk,
  onCancel,
  onOpened,
  width,
}: ModalProps) => {
  const id = `modal_${Date.now()}`;
  const [destroyed, setDestroyed] = useState(false);
  const [loaded, setLoaded] = useState(false);
  const [opened, setOpened] = useState(visible);
  const [maskVisible, setMaskVisible] = useState(visible);
  const [modalVisible, setModalVisible] = useState(visible);
  const modalElmRef = useRef<HTMLDivElement | null>(null);
  const maskElmRef = useRef<HTMLDivElement | null>(null);
  const label = useMemo(() => (title ? title : id), [id, title]);

  const overlayStyle = useSpring({
    config: {
      duration: animate ? 100 : 0,
    },
    onRest: () => {
      if (opened) {
        setModalVisible(true);
      } else {
        if (maskElmRef.current) {
          const level = Number.parseInt(
            maskElmRef.current.getAttribute(`aria-level`) ?? `0`,
            10
          );
          maskElmRef.current.style.zIndex = `9999`;
          maskElmRef.current.setAttribute(`aria-level`, `1`);
          if (level) {
            // get the previous modal and set it back to an active state
            const previousMask: HTMLElement | null = document.querySelector(
              `.${overlayVariants.visible.split(` `)[0]}.inactive[aria-level="${
                level - 1
              }"]`
            );
            if (previousMask) {
              previousMask.classList.remove(`inactive`);
              // reset focus to the currently active modal
              previousMask.focus();
            }
          }
        }

        setMaskVisible(false);

        if (destroyOnClose) {
          setDestroyed(true);
        }

        onCancel?.();
      }
    },
    onStart: () => {
      // set all other modals to an inactive state
      if (opened && maskElmRef.current) {
        Array.from(
          document.querySelectorAll(`.${overlayVariants.visible.split(` `)[0]}`)
        )
          .filter((elm) => elm !== maskElmRef.current)
          .forEach((elm) => elm.classList.add(`inactive`));
      }

      /* 
        make sure that no matter where the modal is placed in the dom that it shows up
        in the correct order in the stack when opened
      */
      const visibleMasks = Array.from(
        document.querySelectorAll(`.${overlayVariants.visible.split(` `)[0]}`)
      );
      if (visibleMasks.length > 1 && maskElmRef.current) {
        maskElmRef.current.setAttribute(`aria-level`, `${visibleMasks.length}`);
        const zIndexArray = visibleMasks.map((mask) => {
          const style = window.getComputedStyle(mask);
          return Number.parseInt(style.zIndex, 10);
        });
        zIndexArray.sort((a, b) => (a > b ? -1 : 1));
        maskElmRef.current.style.zIndex = zIndexArray[0]
          ? `${zIndexArray[0] + 1}`
          : `9999`;
      }
    },
    opacity: opened ? 1 : 0,
  });

  const modalAnimationStyle = useSpring({
    ...(modalVisible
      ? { opacity: 1, scale: 1, y: 0 }
      : { opacity: 0, scale: 0.9, y: 20 }),
    config: {
      duration: animate ? 200 : 0,
    },
    onRest: () => {
      if (modalElmRef.current) {
        modalElmRef.current.focus();
      }
    },
  });

  const closeModal = useCallback(() => {
    setModalVisible(false);
    setOpened(false);
  }, []);

  const onModalClick = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.stopPropagation();
  };

  useEffect(() => {
    setLoaded(true);
  }, []);

  useEffect(() => {
    if (loaded && onOpened && modalVisible) {
      setTimeout(onOpened, animate ? 250 : 0);
    }
  }, [onOpened, loaded, modalVisible, animate]);

  useEffect(() => {
    const maskElement = maskElmRef.current;
    const onEscape = (event: KeyboardEvent) => {
      event.stopPropagation();
      if (event.key === `Escape`) {
        closeModal();
      }
    };

    const onMaskClick = (event: MouseEvent) => {
      if (maskElement === event.target) {
        event.stopPropagation();
        closeModal();
      }
    };

    if (maskElement && escapable) {
      maskElement.addEventListener(`keydown`, onEscape);
    }

    if (maskElement && maskClosable) {
      maskElement.addEventListener(`click`, onMaskClick);
    }

    return () => {
      if (maskElement) {
        maskElement.removeEventListener(`keydown`, onEscape);
        maskElement.removeEventListener(`click`, onMaskClick);
      }
    };
  }, [maskElmRef, loaded, escapable, maskClosable, closeModal]);

  useEffect(() => {
    let firstElm: HTMLElement | null = null;
    let lastElm: HTMLElement | null = null;

    // trap the tab to the inside of the modal
    const onFirstElmKeyDown = (event: KeyboardEvent) => {
      if (event.key === `Tab` && event.shiftKey) {
        event.preventDefault();
        lastElm?.focus();
      }
    };

    const onLastElmKeyDown = (event: KeyboardEvent) => {
      if (event.key === `Tab` && !event.shiftKey) {
        event.preventDefault();
        firstElm?.focus();
      }
    };

    if (loaded && modalElmRef.current && modalVisible) {
      const elms = Array.from(
        modalElmRef.current.querySelectorAll(
          `select, input, textarea, button, a`
        )
      ).filter((elm) => {
        const style = window.getComputedStyle(elm);
        return (
          style.getPropertyValue(`display`) !== `none` &&
          style.getPropertyValue(`visibility`) !== `hidden`
        );
      });
      if (elms.length > 0) {
        firstElm = elms[0] as HTMLElement;
        lastElm = elms[elms.length - 1] as HTMLElement;

        firstElm.addEventListener(`keydown`, onFirstElmKeyDown);
        lastElm.addEventListener('keydown', onLastElmKeyDown);
      }
    }

    return () => {
      if (firstElm && lastElm) {
        firstElm.removeEventListener(`keydown`, onFirstElmKeyDown);
        lastElm.removeEventListener(`keydown`, onLastElmKeyDown);
      }
    };
  }, [modalElmRef, modalVisible, loaded]);

  useEffect(() => {
    if (visible) {
      setOpened(true);
      setDestroyed(false);
      setMaskVisible(true);
    } else {
      closeModal();
    }
  }, [visible, closeModal]);

  // lock body scrolling when modal is visible
  useEffect(() => {
    if (!loaded) return;
    document.body.style.overflowY = opened ? `hidden` : `auto`;
  }, [loaded, opened]);

  if (destroyed || (lazy && !visible)) return null;

  return (
    <Portal>
      <animated.div
        aria-level={1}
        className={`${
          overlayVariants[maskVisible ? 'visible' : 'hidden']
        } active`}
        data-testid="mask"
        ref={maskElmRef}
        style={overlayStyle}
        tabIndex={-1}
      >
        <div className={modalWrapperStyle} style={{ width }}>
          <animated.div
            aria-label={label}
            aria-modal
            className={modalStyle}
            id={id}
            onClick={onModalClick}
            ref={modalElmRef}
            role="dialog"
            style={modalAnimationStyle}
            tabIndex={0}
          >
            {header ?? (
              <div className={headerStyle}>
                {title}
                <button
                  aria-label="close"
                  className={closeButtonStyle}
                  onClick={closeModal}
                  type="button"
                >
                  <Icon icon="carbon:close" />
                </button>
              </div>
            )}
            <div className={bodyStyle}>{children}</div>
            {footer ?? (
              <div className={footerStyle}>
                <SimpleButton onClick={closeModal}>
                  {cancelButtonText}
                </SimpleButton>
                <SimpleButton onClick={onOk} type="primary">
                  {okButtonText}
                </SimpleButton>
              </div>
            )}
          </animated.div>
        </div>
      </animated.div>
    </Portal>
  );
};

export default Modal;
