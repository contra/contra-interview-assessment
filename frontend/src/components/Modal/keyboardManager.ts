
export class KeyboardManager {
  private subscribed = false;

  constructor() {
    this.subscribe();
  }

  private registeredKeys: { [key: string]:  ((e: KeyboardEvent) => void)[] } = {}

  private handleKey(event: KeyboardEvent) {
    let key = event.key;
    const keys = this.registeredKeys[key];

    if (keys && keys.length > 0) {
      const func = keys[keys.length - 1];
      func && func(event);
    }
  }

  subscribe() {
    if (typeof window !== 'undefined' && document && !this.subscribed) {
      this.subscribed = true;
      document.addEventListener('keyup', (e) => this.handleKey(e), false);
    }
  }

  unsubscribe() {
    document.removeEventListener('keyup', (e) => this.handleKey(e), false);
  }

  register(key: string, func: (e: KeyboardEvent) => void) {
    if (!this.subscribed) {
      this.subscribe();
    }
    if (!this.registeredKeys[key]) {
      this.registeredKeys[key] = [];
    }
    this.registeredKeys[key]?.push(func)
  }

  unregister(key: string, func: (e: KeyboardEvent) => void) {
    if (this.registeredKeys[key]) {
      this.registeredKeys[key] = this.registeredKeys[key]?.filter(f => f !== func) || [];
    }
  }
}

export default new KeyboardManager();