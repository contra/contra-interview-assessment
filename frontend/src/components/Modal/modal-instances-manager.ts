import { useEffect, useState } from 'react';

class ModalInstancesManager {
  // eslint-disable-next-line @typescript-eslint/prefer-readonly
  private modals: Map<string, number>;

  public constructor() {
    this.modals = new Map();
  }

  public add(modalId: string): number {
    this.modals.set(modalId, this.modals.size + 1);
    return this.modals.size;
  }

  public remove(modalId: string): void {
    this.modals.delete(modalId);
  }

  public isTopModal(modalId: string | null): boolean {
    if (!modalId) return false;
    return this.modals.get(modalId) === this.modals.size;
  }

  public hasActiveModals(): boolean {
    return Boolean(this.modals.size);
  }
}

export const manager = new ModalInstancesManager();

export const useModalManager = (modalId: string, isOpen?: boolean) => {
  const [index, setIndex] = useState(0);

  useEffect(() => {
    if (isOpen) {
      const modalIndex = manager.add(modalId);
      setIndex(modalIndex);
    }

    // eslint-disable-next-line consistent-return
    return () => {
      manager.remove(modalId);
      setIndex(0);
    };
  }, [isOpen, modalId]);

  return index;
};
