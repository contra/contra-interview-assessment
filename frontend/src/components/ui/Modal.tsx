import { useContext, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { useStore } from 'zustand';
import { ModalContext } from '@/components/context/ModalContext';
import styles from './Modal.module.css';

export type ModalProps = React.PropsWithChildren & {
  handleClose: () => void;
};

export default function Modal({ handleClose, children }: ModalProps) {
  const store = useContext(ModalContext);
  const { addModal, removeModal } = useStore(store);

  useEffect(() => {
    addModal(handleClose);
    return () => removeModal(handleClose);
  }, [addModal, removeModal]);

  return createPortal(
    <div className={styles['modal-background']} onClick={handleClose}>
      <div className={styles['modal']} onClick={(e) => e.stopPropagation()}>
        {children}
      </div>
    </div>,
    document.body
  );
}
