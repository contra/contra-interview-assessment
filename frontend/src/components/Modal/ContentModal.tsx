import React from 'react';
import styled from 'styled-components';
import Modal, { type ModalProps } from '.';

export type ContentModalProps = ModalProps & {
  text: string;
  title: string;
};

const ContentContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 12px;
`;

const Title = styled.h2`
  font-size: ${({ theme: { fontSize } }) => fontSize.larger};
  line-height: ${({ theme: { lineHeight } }) => lineHeight.large};
  color: ${({ theme: { colors } }) => colors.modalTitle};
`;

const Text = styled.p`
  font-size: ${({ theme: { fontSize } }) => fontSize.regular};
  line-height: ${({ theme: { lineHeight } }) => lineHeight.regular};
  color: ${({ theme: { colors } }) => colors.modalText};
`;

const ContentModal: React.FC<ContentModalProps> = ({
  text,
  title,
  children,
  ...baseModalProps
}) => {
  return (
    <Modal {...baseModalProps}>
      <ContentContainer>
        <Title>{title}</Title>
        <Text>{text}</Text>
      </ContentContainer>
      {children}
    </Modal>
  );
};

export default ContentModal;
