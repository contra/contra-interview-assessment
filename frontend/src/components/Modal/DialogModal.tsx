import React from 'react';
import styled from 'styled-components';
import Button from '../Button';
import ContentModal, { type ContentModalProps } from './ContentModal';

type Props = ContentModalProps & {
  cancelLabel: string;
  confirmLabel: string;
  onClickCancel: () => void;
  onClickConfirm: () => void;
  text: string;
  title: string;
};

const ButtonsContainer = styled.div`
  display: flex;
  gap: 16px;
  margin-top: 24px;
  justify-content: flex-end;
`;

const DialogModal: React.FC<Props> = ({
  cancelLabel,
  confirmLabel,
  onClickCancel,
  onClickConfirm,
  ...modalProps
}) => {
  return (
    <ContentModal {...modalProps}>
      <ButtonsContainer>
        <Button onClick={onClickCancel}>{cancelLabel}</Button>
        <Button onClick={onClickConfirm} variant="Secondary">
          {confirmLabel}
        </Button>
      </ButtonsContainer>
    </ContentModal>
  );
};

export default DialogModal;
