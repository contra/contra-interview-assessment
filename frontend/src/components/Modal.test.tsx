import { render, screen } from '@testing-library/react'
import Modal from './Modal';
// eslint-disable-next-line import/no-unassigned-import
import '@testing-library/jest-dom'

const onCloseHandler = jest.fn(() => {});

describe('Modal', () => {

    it('no show', () => {
        const { container } = render(<Modal handleClose={onCloseHandler} show={false} title="Modal">Modal Content</Modal>);

        expect(container).toMatchSnapshot();

        const title = screen.queryByText('Modal');

        expect(onCloseHandler.mock.calls).toHaveLength(0);
        expect(title).toBeFalsy();
    });

    it('show', async () => {
        expect.hasAssertions();

        const { container } = render(<Modal animate={false} handleClose={onCloseHandler} show title="Modal">Modal Content</Modal>);

        expect(container).toMatchSnapshot();

        await screen.findByText('Modal');
        expect(onCloseHandler.mock.calls).toHaveLength(0);
    });

    it('close', async () => {
        expect.hasAssertions();

        render(<Modal animate={false} handleClose={onCloseHandler} show title="Modal">Modal Content</Modal>);

        const close = await screen.findByText('Close');
        close.click();

        expect(onCloseHandler.mock.calls).toHaveLength(1);
    });
});