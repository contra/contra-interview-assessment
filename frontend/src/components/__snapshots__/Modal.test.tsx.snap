// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal no show 1`] = `<div />`;

exports[`Modal show 1`] = `
<div>
  <div
    class="fixed top-0 left-0 w-full h-full"
  >
    <div
      aria-hidden="true"
      class="absolute top-0 left-0 w-full h-full bg-slate-300/80"
      role="button"
      style="z-index: 1000;"
      tabindex="0"
    />
    <div
      aria-label="Modal"
      class="absolute w-full sm:w-[400px] sm:-ml-[200px] bg-white sm:rounded-xl shadow-xl top-1/4 sm:left-1/2 mt-0"
      role="dialog"
      style="z-index: 1001;"
      tabindex="-1"
    >
      <h2
        class="text-center bg-violet-700 sm:rounded-t-xl text-white py-1"
      >
        Modal
      </h2>
      <div
        class="py-8 px-8 space-y-2 sm:py-4 flex items-center flex-col sm:space-y-0 sm:space-x-6"
      >
        Modal Content
      </div>
      <div
        class="text-center mb-4"
      >
        <button
          class="transition ease-in-out duration-300 active:bg-black hover:bg-violet-400 hover:shadow-xl focus:bg-black font-medium text-lg min-w-[130px] px-4 h-14 rounded-full shadow outline-none bg-violet-700 text-white"
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
